### PR TITLE
Add webhook for ClusterBootstrap and ClusterBootstrapTemplate

### LINKED ISSUE #1689: Webhook for BootstrapClusterTemplate/BootstrapCluster
Webhook for validating clusterbootstraptemplate/clusterbootstrap

All the fields in a ClusterBootStrapTemplate are immutable

ClusterBootstrap webook

For cni, cpi, csi, kapp once created

1.   we won’t allow packageRef’s to be downgraded or change the package from something like calico to antrea
2.  We can start with disallowing change of apiVersion and Kind. In the future we can relax this
3.  Can change inline or secret to whatever

For Additional packages that are created
1.   no deletion of a package allowed, in the future we can consider relaxing this
2.  Can bump package version
3.  Not allowed to change apiVersion and Kind for provider
4.  Can change inline or secret to whatever

### PR DESCRIPTION
**Pull request by blc1996**
### What this PR does / why we need it

We need to add webhook for ClusterBootstrap and ClusterBootstrapTemplate to enforce some validation rules

### Which issue(s) this PR fixes

Fixes #1689 

### Release note

```release-note
package-based-lcm: Adds validation webhook for ClusterBoostrap API
```

### PR Checklist

- [x] Squash the commits into one or a small number of logical commits
      
- [x] Use good commit [messages](https://github.com/vmware-tanzu/tanzu-framework/blob/main/CONTRIBUTING.md)
- [x] Ensure PR contains terms all contributors can understand and links all contributors can access

### CODE REVIEW

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+		clusterBootstrap.Name, allErrs)
+}
+
+func (webhook *ClusterBootstrap) ValidateClusterBootstrapPackage(ctx context.Context, pkg *runv1alpha3.ClusterBootstrapPackage, namespace string, fldPath *field.Path) *field.Error {
+	if pkg == nil {
+		return field.Invalid(fldPath, pkg, "package can't be nil")
+	}
+
+	// The package URL in refName must be valid
+	if _, err := name.ParseReference(pkg.RefName); err != nil {
```
Comment (blc1996) [inline addons/webhooks/clusterbootstrap_webhook.go:129]: AI: update this

File: addons/test/testutil/test_helpers.go [hunk 2]
```diff
@@ -42,6 +42,63 @@ import (
 	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/webhooks"
 )
 
+type WebhookCertificatesDetails struct {
```
Comment (shivaani0505) [inline addons/test/testutil/test_helpers.go:45]: @blc1996 , My PR is already merged, so I am wondering why this change is showing up here in your diff

Comment (blc1996) [reply addons/test/testutil/test_helpers.go:45]: I move it up a bit

File: config/crd/patches/cainjection_in_clusterbootstraps.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# The following patch adds a directive for certmanager to inject CA into the CRD
```
Comment (shivaani0505) [inline config/crd/patches/cainjection_in_clusterbootstraps.yaml:1]: Are we adding the cainjection files back. I also had added them but had to remove them. So we can be consistent here for all the controllers

File: config/crd/patches/cainjection_in_clusterbootstraptemplates.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# The following patch adds a directive for certmanager to inject CA into the CRD
+# CRD conversion requires k8s 1.13 or later.
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (shivaani0505) [inline config/crd/patches/cainjection_in_clusterbootstraptemplates.yaml:3]: same as above

File: addons/controllers/calicoconfig_controller_test.go [hunk 5]
```diff
@@ -162,7 +162,7 @@ var _ = Describe("CalicoConfig Reconciler and Webhooks", func() {
 			Expect(k8sClient.Get(ctx, key, config)).To(Succeed())
 
 			By("Trying to update the immutable Namespace field in Calico Spec")
-			config.Spec.Namespace = "default"
+			config.Spec.Namespace = key.Namespace
```
Comment (blc1996) [inline addons/controllers/calicoconfig_controller_test.go:165]: linter complains "default" can be a constant

File: config/crd/patches/webhook_in_clusterbootstraps.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+# The following patch enables conversion webhook for CRD
+# CRD conversion requires k8s 1.13 or later.
```
Comment (shivaani0505) [inline config/crd/patches/webhook_in_clusterbootstraps.yaml:2]: same for webhook_in_* files , we need to decide if we include them for all controllers or remove them.

File: config/samples/run_v1alpha3_clusterbootstrap.yaml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+apiVersion: run.tanzu.vmware.com/v1alpha3
+kind: ClusterBootstrap
+metadata:
+  name: clusterbootstrap-sample
+spec:
+  # TODO(user): Add fields here
```
Comment (shivaani0505) [inline config/samples/run_v1alpha3_clusterbootstrap.yaml:6]: Add sample values for `spec`

File: config/samples/run_v1alpha3_clusterbootstraptemplate.yaml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+apiVersion: run.tanzu.vmware.com/v1alpha3
+kind: ClusterBootstrapTemplate
+metadata:
+  name: clusterbootstraptemplate-sample
+spec:
+  # TODO(user): Add fields here
```
Comment (shivaani0505) [inline config/samples/run_v1alpha3_clusterbootstraptemplate.yaml:6]: Add sample values for spec

File: config/crd/patches/webhook_in_clusterbootstraps.yaml (hunk 6 above)
Comment (blc1996) [reply config/crd/patches/webhook_in_clusterbootstraps.yaml]: Removed

File: config/crd/patches/cainjection_in_clusterbootstraptemplates.yaml (hunk 4 above)
Comment (blc1996) [reply config/crd/patches/cainjection_in_clusterbootstraptemplates.yaml]: removed

File: config/crd/patches/cainjection_in_clusterbootstraps.yaml (hunk 3 above)
Comment (blc1996) [reply config/crd/patches/cainjection_in_clusterbootstraps.yaml]: removed

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+	if len(allErrs) == 0 {
+		return nil
+	}
+
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "run.tanzu.vmware.com", Kind: "ClusterBootstrap"},
+		newClusterBootstrap.Name, allErrs)
+}
+
+func (wh *ClusterBootstrap) ValidateClusterBootstrapPackageUpdate(ctx context.Context, oldPkg, newPkg *runv1alpha3.ClusterBootstrapPackage, fldPath *field.Path) *field.Error {
```
Comment (vijaykatam) [inline addons/webhooks/clusterbootstrap_webhook.go:332]: Question: why is the function exported?

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 10]
```diff
@@ -0,0 +1,413 @@
...
+		// We don't allow changes to APIGroup and Kind of providerRef
+		if oldValuesFrom.ProviderRef != nil && newValuesFrom.ProviderRef != nil {
+			if *oldValuesFrom.ProviderRef.APIGroup != *newValuesFrom.ProviderRef.APIGroup ||
+				oldValuesFrom.ProviderRef.Kind != newValuesFrom.ProviderRef.Kind {
+				return field.Invalid(fldPath.Child("ProviderRef"), newValuesFrom.ProviderRef, "change to Group and Kind in ProviderRef is not allowed")
+			}
+		}
+
+		// Users can't switch from ProviderRef to secretRef/inline
+		if oldValuesFrom.ProviderRef != nil && newValuesFrom.ProviderRef == nil {
```
Comment (vijaykatam) [inline addons/webhooks/clusterbootstrap_webhook.go:402]: This is checking if new ProviderRef is nil and not what the comment states.

File: addons/controllers/clusterbootstrap_controller_test.go [hunk 11]
```diff
@@ -463,6 +502,103 @@ var _ = Describe("ClusterBootstrap Reconciler", func() {
...
+					mutateClusterBootstrap = clusterBootstrap.DeepCopy()
+					mutateClusterBootstrap.Spec.CNI.RefName = "antrea.tanzu.vmware.com.0.13.3--vmware.1-tkg.1"
+					err = k8sClient.Update(ctx, mutateClusterBootstrap)
+					Expect(strings.Contains(err.Error(), "package downgrade is not allowed")).To(BeTrue())
+
+					// Additional package can't be removed
+					mutateClusterBootstrap = clusterBootstrap.DeepCopy()
+					mutateClusterBootstrap.Spec.AdditionalPackages = mutateClusterBootstrap.Spec.AdditionalPackages[:len(mutateClusterBootstrap.Spec.AdditionalPackages)-1]
+					err = k8sClient.Update(ctx, mutateClusterBootstrap)
+					Expect(strings.Contains(err.Error(), "missing updated additional package")).To(BeTrue())
```
Comment (vijaykatam) [inline addons/controllers/clusterbootstrap_controller_test.go:562]: Could we add a testcase for changing provider to secret and inline?

**Review by vijaykatam — COMMENTED** [review]
I think this is fairly close. Code comments are helpful especially around validating upgrade.

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 12]
```diff
@@ -0,0 +1,413 @@
...
+	}
+
+	// TODO: validation for inline manifests? No-op for now
+
+	return nil
+}
+
+// getGVR returns a GroupVersionResource for a GroupKind
+func (wh *ClusterBootstrap) getGVR(gk schema.GroupKind) (*schema.GroupVersionResource, error) {
+	apiResourceList, err := wh.cachedDiscoveryClient.ServerPreferredResources()
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go:202]: Added GVR cache, used SIGTERM and SIGINT handler as context

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+			Name:      valuesFrom.SecretRef,
+			Namespace: clusterBootstrapNamespace,
+		}
+		err := wh.Client.Get(ctx, key, valueSecret)
+		if err != nil {
+			return field.Invalid(fldPath.Child("SecretRef"), valuesFrom.SecretRef, err.Error())
+		}
+	}
+
+	// TODO: validation for inline manifests? No-op for now
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go:195]: Removed comment

File: addons/webhooks/clusterbootstraptemplate_webhook.go [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+		For(&runv1alpha3.ClusterBootstrapTemplate{}).
+		WithDefaulter(wh).
+		WithValidator(wh).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-run-tanzu-vmware-com-v1alpha3-clusterbootstraptemplate,mutating=true,failurePolicy=fail,groups=run.tanzu.vmware.com,resources=clusterbootstraptemplates,verbs=create;update,versions=v1alpha3,name=mclusterbootstraptemplate.kb.io
+//+kubebuilder:webhook:verbs=create;update,path=/validate-run-tanzu-vmware-com-v1alpha3-clusterbootstraptemplate,mutating=false,failurePolicy=fail,groups=run.tanzu.vmware.com,resources=clusterbootstraptemplates,versions=v1alpha3,name=vclusterbootstraptemplate.kb.io
+
+var _ webhook.CustomDefaulter = &ClusterBootstrapTemplate{}
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstraptemplate_webhook.go:42]: Removed mutating webook for CBT

File: addons/webhooks/clusterbootstraptemplate_webhook.go [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected a Cluster but got a %T", obj))
+	}
+	clusterbootstraplog.Info("validate create", "name", clusterBootstrapTemplate.Name)
+
+	var allErrs field.ErrorList
+	// Iterating one by one because we need field info for getFieldPath, and some core packages can be nil
+
+	if clusterBootstrapTemplate.Namespace != wh.SystemNamespace {
+		nsErrMsg := fmt.Sprintf("ClusterBootstrapTemplate can only be created in tanzu system namespace %s", wh.SystemNamespace)
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstraptemplate_webhook.go:66]: Updated

File: addons/webhooks/clusterbootstraptemplate_webhook.go [hunk 16]
```diff
@@ -0,0 +1,159 @@
...
+	if (pkg.ValuesFrom.ProviderRef != nil && pkg.ValuesFrom.SecretRef != "") ||
+		(pkg.ValuesFrom.ProviderRef != nil && pkg.ValuesFrom.Inline != "") ||
+		(pkg.ValuesFrom.SecretRef != "" && pkg.ValuesFrom.Inline != "") {
+		return field.Invalid(fldPath.Child("valuesFrom"), pkg.ValuesFrom, "valuesFrom can't have more than one non-null subfield")
+	}
+
+	// Will not check if provider config CRs, secretRefs or package CRs exist in the cluster for clusterBootstrapTemplate
+	// As there is no ordering guaranteed in creating TKR resources, i.e. CBT can be created before other resources
+
+	if pkg.ValuesFrom.ProviderRef != nil && pkg.ValuesFrom.ProviderRef.APIGroup == nil {
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstraptemplate_webhook.go:128]: Correct me if I'm wrong. Looks like it's a pointer type and it's optional.

```
	// +optional
	APIGroup *string `json:"apiGroup" protobuf:"bytes,1,opt,name=apiGroup"`
```

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+	if clusterBootstrap.Spec.CPI != nil {
+		if err := wh.validateClusterBootstrapPackage(ctx, clusterBootstrap.Spec.CPI, clusterBootstrap.Namespace, getFieldPath("cpi")); err != nil {
+			allErrs = append(allErrs, err)
+		}
+	}
+
+	if clusterBootstrap.Spec.AdditionalPackages != nil {
+		// validate additional packages
+		for _, pkg := range clusterBootstrap.Spec.AdditionalPackages {
+			if err := wh.validateClusterBootstrapPackage(ctx, pkg, clusterBootstrap.Namespace, getFieldPath("additionalPackages")); err != nil {
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go:119]: Yes, I also noticed it when I was implementing it. Since it's an array, looks like the only thing we can do is to include the index.

Added the index in the fieldPath

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+	//	b. We can start with disallowing change of apiVersion and Kind. In the future we can relax this
+	//	c. Can change inline or secret to whatever
+	//
+	//	2. For Additional packages that are created
+	//	a. no deletion of a package allowed, in the future we can consider relaxing this
+	//	b. Can bump package version
+	//  b. Not allowed to change apiVersion and Kind for provider
+	//	c. Can change inline or secret to whatever
+
+	if oldPkg == nil || newPkg == nil {
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go:344]: Yes, the nil check for CSI and CPI are already handled in `validateOptionalCorePackageUpdate` under `ValidateUpdate`.

File: addons/webhooks/clusterbootstrap_webhook.go (hunk 9 above)
Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go:332]: Fixed

File: addons/webhooks/clusterbootstraptemplate_webhook.go [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+		return field.Invalid(fldPath, pkg, "package can't be nil")
+	}
+
+	// valuesFrom can be nil
+	if pkg.ValuesFrom == nil {
+		return nil
+	}
+
+	// Currently, we don't allow more than one field from valuesFrom to be present
+	if (pkg.ValuesFrom.ProviderRef != nil && pkg.ValuesFrom.SecretRef != "") ||
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstraptemplate_webhook.go:119]: Fixed

File: addons/controllers/clusterbootstrap_controller_test.go (hunk 11 above)
Comment (blc1996) [reply addons/controllers/clusterbootstrap_controller_test.go:585]: Added

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 20]
```diff
@@ -0,0 +1,413 @@
...
+func (wh *ClusterBootstrap) ValidateClusterBootstrapPackageUpdate(ctx context.Context, oldPkg, newPkg *runv1alpha3.ClusterBootstrapPackage, fldPath *field.Path) *field.Error {
+	//	1. For cni, cpi, csi, kapp once created
+	//	a. we won’t allow packageRef’s to be downgraded or change the package from something like calico to antrea
+	//	b. We can start with disallowing change of apiVersion and Kind. In the future we can relax this
+	//	c. Can change inline or secret to whatever
+	//
+	//	2. For Additional packages that are created
+	//	a. no deletion of a package allowed, in the future we can consider relaxing this
+	//	b. Can bump package version
+	//  b. Not allowed to change apiVersion and Kind for provider
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go]: Updated

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+
+	return apierrors.NewInvalid(
+		schema.GroupKind{Group: "run.tanzu.vmware.com", Kind: "ClusterBootstrap"},
+		newClusterBootstrap.Name, allErrs)
+}
+
+func (wh *ClusterBootstrap) ValidateClusterBootstrapPackageUpdate(ctx context.Context, oldPkg, newPkg *runv1alpha3.ClusterBootstrapPackage, fldPath *field.Path) *field.Error {
+	//	1. For cni, cpi, csi, kapp once created
+	//	a. we won’t allow packageRef’s to be downgraded or change the package from something like calico to antrea
+	//	b. We can start with disallowing change of apiVersion and Kind. In the future we can relax this
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go]: Updated

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+
+// validateValuesFrom validates content of valuesFrom
+func (wh *ClusterBootstrap) validateValuesFrom(ctx context.Context, valuesFrom *runv1alpha3.ValuesFrom, clusterBootstrapNamespace string, fldPath *field.Path) *field.Error {
+	// valuesFrom can be nil
+	if valuesFrom == nil {
+		return nil
+	}
+
+	// Currently, we don't allow more than one field from valuesFrom to be present
+	if (valuesFrom.ProviderRef != nil && valuesFrom.SecretRef != "") ||
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go]: Updated

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 23]
```diff
@@ -0,0 +1,413 @@
...
+		return field.Invalid(fldPath.Child("refName"), newPkg.RefName, retErr.Error())
+	}
+	oldPkgSemver, err := versions.NewRelaxedSemver(oldPackageVersion)
+	if err != nil {
+		retErr := errors.Wrap(err, "old package version is invalid")
+		return field.Invalid(fldPath.Child("refName"), oldPkg.RefName, retErr.Error())
+	}
+	if newPkgSemver.Compare(oldPkgSemver.Version) == -1 {
+		// package downgrade is not allowed
+		return field.Invalid(fldPath.Child("refName"), newPkg.RefName, "package downgrade is not allowed")
```
[the remark this answers is not in the document]
Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go]: Printed out the original and updated version

File: addons/webhooks/clusterbootstrap_webhook.go (hunk 10 above)
Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go:434]: Updated

File: addons/webhooks/clusterbootstrap_webhook.go [hunk 24]
```diff
@@ -0,0 +1,442 @@
...
+	if err := wh.validateOptionalCorePackageUpdate(ctx, oldClusterBootstrap.Spec.CSI, newClusterBootstrap.Spec.CSI, namespace, getFieldPath("csi")); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	if err := wh.validateOptionalCorePackageUpdate(ctx, oldClusterBootstrap.Spec.CPI, newClusterBootstrap.Spec.CPI, namespace, getFieldPath("cpi")); err != nil {
+		allErrs = append(allErrs, err)
+	}
+
+	// validate new additional packages
+	addtionalPkgFldPath := getFieldPath("additionalPackages")
```
Comment (vijaykatam) [inline addons/webhooks/clusterbootstrap_webhook.go:310]: Does this need to be changed to `getFieldPath("additionalPackages").Child(string(rune(idx)))` like above?

Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go:325]: Thanks for pointing out. Changed for all the additional packages.

Comment (vijaykatam) [reply addons/webhooks/clusterbootstrap_webhook.go:325]: Is there a test to exercise this flow?

Comment (blc1996) [reply addons/webhooks/clusterbootstrap_webhook.go:325]: There is a testcase for `// Additional package can't be removed`, all other logics in validating spec should be the same as core packages, as they are calling the same functions.

File: config/samples/run_v1alpha3_clusterbootstrap.yaml (hunk 7 above)
Comment (vijaykatam) [reply config/samples/run_v1alpha3_clusterbootstrap.yaml:6]: Could we remove this sample if not used anywhere? Same comment for below.

File: packages/management/addons-manager-v2/bundle/config/upstream/webhook-manifests.lib.yaml [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: tanzu-addons-manager-webhook-service
+  namespace: #@ data.values.tanzuAddonsManager.namespace
+spec:
+  ports:
+  - port: 443
+    targetPort: 9453
```
Comment (vijaykatam) [inline packages/management/addons-manager-v2/bundle/config/upstream/webhook-manifests.lib.yaml:13]: update to 9865 (see https://github.com/vmware-tanzu/tanzu-framework/pull/2138/files)

File: addons/webhooks/clusterbootstrap_webhook.go (hunk 24 above)
Comment (vijaykatam) [reply addons/webhooks/clusterbootstrap_webhook.go:318]: Had an offline sync. fieldPath is used in error messages and not for the actual object.

**Review by vijaykatam — APPROVED** [review]
Nice work! LGTM
